gem/refactorise.py

```diff
@@ -112,6 +116,175 @@ def product(*args):
+
+        # Form IndexSum's from each monomial group
+        for (_, sum_indices), monomials in iteritems(monomial_group):
+            all_atomics = [m[0] for m in monomials]
+            all_rest = [m[1] for m in monomials]
+            if len(all_atomics) == 1:
+                # Just one term, add to indexsums directly
+                indexsums.append(fast_sum_factorise(sum_indices, all_atomics[0] + (all_rest[0],)))
+            else:
+                # Form products for each monomial
+                products = [associate_product(atomics + (_rest,))[0] for atomics, _rest in zip(all_atomics, all_rest)]
+                indexsums.append(IndexSum(associate_sum(products)[0], sum_indices))
+
+        return associate_sum(indexsums)[0]
+
+    def find_optimal_atomics(self, sum_indices):
+        sum_indices_set, _ = sum_indices
+        index = count()
+        atomic_index = OrderedDict()  # Atomic gem node -> int
+        connections = []
+        # add connections (list of lists)
+        for (_sum_indices, _), (_, atomics) in iteritems(self.ordering):
+            if _sum_indices == sum_indices_set:
+                connection = []
+                for atomic in atomics:
+                    if atomic not in atomic_index:
+                        atomic_index[atomic] = next(index)
+                    connection.append(atomic_index[atomic])
+                connections.append(tuple(connection))
+
+        if len(atomic_index) == 0:
+            return ((), ())
+        if len(atomic_index) == 1:
+            return ((list(atomic_index.keys())[0], ), ())
+
+        # set up the ILP
+        import pulp as ilp
+        ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+        ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+        # Objective function
+        # Minimise number of factors to pull. If same number, favour factor with larger extent
+        big = 10000000  # some arbitrary big number
+        ilp_prob += ilp.lpSum(ilp_var[index] * (big - self.argument_indices_extent(atomic)) for atomic, index in iteritems(atomic_index))
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+        ilp_prob.solve()
+        if ilp_prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 1]
+        other_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 0]
+        optimal_atomics = sorted(optimal_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        other_atomics = sorted(other_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        return (tuple(optimal_atomics), tuple(other_atomics))
```

Comment:
Again, this can now move to be a free function in `coffee_mode.py`.  That way, I also (if I'm not wanting coffee), I don't have to import YAPP (pulp in this case).